# README

## processed/ukbb/helperfiles/ukbfields.tsv

* Generated by ./scripts/02-UKBB_sampleQC.R
* Decoded UKBB field names to field ID, visit, and number, including the definition of field IDs. 

## processed/ukbb/helperfiles/discSex_exclusions.tsv

* Generated by ./scripts/02-UKBB_sampleQC.R
* The list of sample ids excluded from the analysis based on the discordant sex criteria.

## processed/ukbb/helperfiles/recommended_exclusions.tsv

* Generated by ./scripts/02-UKBB_sampleQC.R
* The list of sample ids excluded based on UK Biobank recommendation, considering heterozygosity and missingness.

## processed/ukbb/helperfiles/sampleIDs_passedQC.tsv

* Generated by ./scripts/02-UKBB_sampleQC.R
* The list of sample IDs that passed the QC process. 

## processed/traits_clean/traitData_baseline(.rds|.tsv)

* Generated by ./scripts/03-prepTraitData.R
* Includes the following information for 484,598 samples that passed the QC.

```{r}
> colnames(traitbaseline)
 [1] "eid"                                         
 [2] "Sex"                                         
 [3] "Standing height"                             
 [4] "Number of self-reported cancers"             
 [5] "Number of self-reported non-cancer illnesses"
 [6] "Number of operations, self-reported"         
 [7] "Number of treatments/medications taken"      
 [8] "Sleep duration"                              
 [9] "Facial ageing"                               
[10] "Maternal smoking around birth"               
[11] "Father's age at death"                       
[12] "Overall health rating"                       
[13] "Age when periods started (menarche)"         
[14] "Mother's age at death"                       
[15] "Age at menopause (last menstrual period)"    
[16] "Non-accidental death in close genetic family"
[17] "Health satisfaction"                         
[18] "Smoking status"                              
[19] "Alcohol drinker status"                      
[20] "Weight"                                      
[21] "Age when attended assessment centre"         
[22] "Age at recruitment"                          
[23] "Age at death"
```

## processed/traits_clean/SRDisease_baseline(.rds|.tsv)

* Generated by ./scripts/03-prepTraitData.R
* Includes the following information for 362,772 samples that passed the QC and has at least one disease (note that the number of entries (904,035) is higher than the number of samples as one sample can have more than one disease).

```{r}
> colnames(SRdiseasebaseline)
[1] "eid"        "diseaseID" 
[3] "Disease"    "node_id"   
[5] "parent_id"  "selectable"
[7] "Age"
```

## processed/traits_clean/SRcancer_baseline(.rds|.tsv)

* Generated by ./scripts/03-prepTraitData.R
* Includes the following information for 39,910 samples that passed the QC and has at least one disease (note that the number of entries (42,852) is higher than the number of samples as one sample can have more than one type of cancer).

```{r}
> colnames(SRcancerbaseline)
[1] "eid"        "cancerID"   "Cancer"     "node_id"    "parent_id"  
[6] "selectable" "Age"
```

## processed/traits_clean/SRDisease_baseline_propagated.rds

* Generated by ./scripts/03-prepTraitData.R
* The self-reported diseases in UK Biobank are not independent, but rather are organized in a hierarchical manner. Using the relationship information between diseases, we propagate disease-participant associations. For example, if a person reports having “essential hypertension”, we also annotate that person with “hypertension”, and “cardiovascular disease”.
* Includes the following information for 362,772 samples that passed the QC and has at least one disease (note that the number of entries (2,242,698) is higher than the number of samples as one sample can have more than one disease).

```{r}
> colnames(SRdiseasebaseline)
[1] "eid"        "diseaseID" 
[3] "Disease"    "node_id"   
[5] "parent_id"  "selectable"
[7] "Age"
```
* Age: Using the minimum age at diagnosis for all the diseases that are child term for a particular disease, we re-defined age at diagnosis information.

## processed/traits_clean/traitData_baseline_additions.rds

* Generated by ./scripts/03-prepTraitData.R
* Includes the following information for 484,598 samples that passed the QC.

```{r}
> colnames(traits)
 [1] "eid"                                          "nSRdiseaseProp"                              
 [3] "nSRcancer"                                    "nSRdisease"                                  
 [5] "Sex"                                          "Standing height"                             
 [7] "Number of self-reported cancers"              "Number of self-reported non-cancer illnesses"
 [9] "Number of operations, self-reported"          "Number of treatments/medications taken"      
[11] "Sleep duration"                               "Facial ageing"                               
[13] "Maternal smoking around birth"                "Father's age at death"                       
[15] "Overall health rating"                        "Age when periods started (menarche)"         
[17] "Mother's age at death"                        "Age at menopause (last menstrual period)"    
[19] "Non-accidental death in close genetic family" "Health satisfaction"                         
[21] "Smoking status"                               "Alcohol drinker status"                      
[23] "Weight"                                       "Age when attended assessment centre"         
[25] "Age at recruitment"                           "Age at death"                                
[27] "BMI"                                          "Parent Min Age at Death"                     
[29] "Parent Max Age at Death"                      "Parent Avg Age at Death" 
```

* nSRdisease: The number of unique self-reported non-cancer illnesses each participant informed in the baseline recruitment. 
* nSRdiseaseProp: The number of unique self-reported diseases each participant has after the data propagation.
* nSRcancer: The number of unique self-reported cancers each participant informed in the baseline recruitment.
* BMI: Using the columns for ‘Weight’ and ‘Standing height’ we calculated BMI based on the following formula: “Weight / (Standing Height / 100)^2”
* Parent Min Age at Death: The age at which the first deceased parent died. If none of the parents died, or if the data is unavailable, this column is set to be NA. If only one parent died, we use the corresponding age as the minimum. 
* Parent Max Age at Death: The age at which the second deceased parent died. If none of the parents died, or if the data is unavailable, this column is set to be NA. If only one parent died, we use the corresponding age as the maximum. 
* Parent Avg Age at Death: The age at which the first deceased parent died. If none of the parents died, or if the data is unavailable, this column is set to be NA. If only one parent died, we use the corresponding age as the average. 
  
## processed/ukbb/gwas/fam4bolt/*.fam

* Generated by ./scripts/prepGWAS/01-fix_fam_files.R
* UKBiobank data has batch in its phenotype column. But this field is coded with characters, whereas BOLT-LMM requires fam files to consist of all numeric values. Using the script we convert the sixth column to numeric equivalents.

## processed/ukbb/gwas/remove/*.fam

* Generated by ./scripts/prepGWAS/01-removeFiles4bolt.R
* BOLT-LMM accepts a list of individuals to be removed from the analysis as an input. These files are called ‘remove’ files and are in the FAM format. We prepared these files for i) withdrawn samples (withdrawn.fam - n = 51), ii) samples that failed the quality control (sampleQC_exc.fam - n = 3,779), iii) samples that have information in PLINK files but lack BGEN files (inplink_notin_bgen.fam - n = 968), and in order to run GWAS using only females or males, we also prepared files including iv) males (males.fam - n = 221,840) and v) females (famelas.fam - n = 262,758).

## raw/ukbb/genotypes/

* Generated by ./scripts/prepGWAS/03-unzip4plink.sh.
* Plink does not work well with zipped files. So we created a local copy of the genotype files. Here we also include fam files under ./data/processed/ukbb/gwas/fam4bolt since it is easier to run code when all bed bim fam files are together.

## processed/ukbb/gwas/hwe

* Generated by ./scripts/prepGWAS/04-calcSNPstats.sh
* Hardy-Weinberg equilibrium stats calculated using PLINK.

## processed/ukbb/gwas/freq

* Generated by ./scripts/prepGWAS/04-calcSNPstats.sh
* Allele frequencies calculated using PLINK.

## processed/ukbb/gwas/exclude/hwe.txt

* Generated by ./scripts/prepGWAS/05-excludeFiles4bolt.R
* SNPs with a p<=1e-6 (n = 202,473).

## processed/ukbb/gwas/exclude/maf.txt

* Generated by ./scripts/prepGWAS/05-excludeFiles4bolt.R
* SNPs with MAF < 0.01 (n = 314,697).

## processed/traits_clean/SRdisease_prop_prevdf.rds

* Generated by ./scripts/05-DiseaseSet.R
* For all 472 self reported disease in UK Biobank, includes the following information:
    * Disease: Name of the disease
    * Female: Number of females with the disease
    * Male: Number of males with the disease
    * fPrev: Disease frequency among females (N in 1,000)
    * mPrev: Disease frequency among males (N in 1,000)
    * nCases: Total number of samples with the disease
    * fCase: Overall disease frequency (N in 1,000)
    
## processed/traits_clean/SRdiseaseSet.rds

* Generated by ./scripts/05-DiseaseSet.R
* A subset of processed/traits_clean/SRdisease_prop_prevdf.rds including only the diseases to be analysed by GWAS. This set includes 116 that are not sex specific, common, and with at least 2,000 cases in UK Biobank.
